cadence/server/api_actions.go

```diff
@@ -0,0 +1,197 @@
+// api_actions.go
+// Repeatable actions used by API functions. Mostly database and audio software calls.
+// No direct responses to clients here.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/Jeffail/gabs"
+	"github.com/kenellorando/clog"
+)
+
+// Song file metadata object
+type SongData struct {
+	ID     int
+	Artist string
+	Title  string
+	Album  string
+	Genre  string
+	Year   int
+}
+
+// Takes a query string to search the database.
+// Returns a slice of SongData of songs ordered by relevance.
+func searchByQuery(query string) (queryResults []SongData, err error) {
+	clog.Info("searchByQuery", fmt.Sprintf("Searching database for query: '%v'", query))
+
+	selectWhereStatement := fmt.Sprintf("SELECT \"rowid\", \"artist\", \"title\",\"album\", \"genre\", \"year\" FROM %s ", c.MetadataTable) + "WHERE artist LIKE $1 OR title LIKE $2 ORDER BY rank"
```

Comment:
Well, it's definitely much nicer than the mess I wrote to get good results out of Levenshtein distance in 4.x...

How good's the rank result?

Reply:
This uses https://www.sqlite.org/fts5.html, which comes from sqlite_fts5 (full text search 5) installed when the container is built. I'm not sure how good the rankings are. I'll have to upload all of my covers and copies of OMR again to my prototype instance.